File: autocorrelation.py
import numpy as np
import matplotlib.pyplot as plt
from scipy.signal import butter, filtfilt
from scipy import stats

# Load Data
filename = 'DLS_files/0.75micro_100dil/30deg.prn'  # Change to your file's path if necessary
data = np.loadtxt(filename, delimiter='\t', skiprows=3)  # Skips first 3 rows
t = data[:, 0]  # Time
voltage = data[:, 1]  # Voltage

fs = 1000 / (t[1] - t[0])  # Sampling frequency
f0 = 100  # Frequency to remove (50 Hz)

# Create the bandstop filter
b, a = butter(2, [f0 - 1, f0 + 1], btype='bandstop', fs=fs)

# Apply the bandstop filter to the voltage signal
voltage_filtered = filtfilt(b, a, voltage)

# Get the number of samples (N)
N = len(voltage_filtered)

# Calculate Autocorrelation function (ACF)
ac = np.zeros(N-1)  # Prepare the array for autocorrelation
m = np.mean(voltage_filtered)  # Mean of the data

# Compute the autocorrelation
for j in range(1, N):
    ac[j-1] = np.mean((voltage_filtered[0:N-j] - m) * (voltage_filtered[j:N] - m))

# Exponential decay fitting using a linear fit on log(ACF)
rangemax = 30  # Limit of data to use for fitting

# Perform the linear fit on the log of the autocorrelation for the first 'rangemax' points
x = np.arange(rangemax)
y = np.log(ac[:rangemax])

# Linear fit: y = m * x + b
slope, intercept, r_value, p_value, std_err = stats.linregress(x, y)

# Characteristic time of exponential decay
tau = -1 / slope
print(f"Tau (characteristic time): {tau}")

# Plot the autocorrelation function with the exponential fit
plt.figure(figsize=(10, 6))

# Plot the first 'rangemax' data points
plt.semilogy(x, ac[:rangemax], 'ro', label='Exp data')
plt.plot(x, np.exp(slope * x + intercept), linewidth=2, label='Fit - exp(-t/τ)')

# Display the value of tau on the graph
plt.text(2, 0.5, f"$\\tau = {tau:.3f}$", fontsize=12, color='blue')

plt.xlabel(r'Time [ms]', fontsize=14)
plt.ylabel(r'$C(t)\ [V^2]$', fontsize=14)
plt.legend(loc='best', fontsize=14)
plt.title(r'Correlation function with Exponential Fit (Semi-Log scale)', fontsize=14)
plt.show()

# Plot the ACF with the exponential fit in linear scale for a larger range
rangemax = 500  # Increase range for linear plot

plt.figure(figsize=(10, 6))
# Use 't' for the time values in the plot instead of np.arange(rangemax)
plt.plot(t[:rangemax], ac[:rangemax], 'ro', markersize=3, label='Exp data')
plt.plot(t[:rangemax], np.exp(slope * t[:rangemax] + intercept), linewidth=2, label='Fit - exp(-t/τ)')

# Display the value of tau on the graph
plt.text(0.1, 0.2, f"$\\tau = {tau:.3f}$", fontsize=12, color='blue')

plt.xlabel(r'Time [ms]', fontsize=14)
plt.ylabel(r'$C(t)\ [V^2]$', fontsize=14)
plt.legend(loc='best', fontsize=14)
plt.title(r'Correlation function with Exponential Fit (Linear scale)', fontsize=14)
plt.show()

File: build_DLS/autocorrelation.m
% Calculates Autocorrelation function from given data. 
% Data assumed to be in pairs (time, value).

%% LOAD DATA

pathdir= 'C:\Users\Part 2 Users\Documents\DLS\E2b\DLS_files';  %set the directory data name
cd(pathdir);

% Import data
filename= 'data2.prn';    %% name of your file
A = importdata(filename,'\t',3);
data = A.data(:,2);     %   voltage
t = A.data(:,1);        %   time steps

% Get no. of samples.  N was previously limited to 10^6 
N=length(data);

%% calculate autocorrelation function

ac= zeros(1,N-1);     %% prepare the variable
m=mean(data);       %% mean of the data
%%%% effective calculation of the autocorrelation
for j= 1:N-1
    ac(j)= mean((data(1:end-j)-m).*(data(j+1:end)-m));
end

%%%%% plot the autocorrelation function in linear and semilogy scale
%figure(1)
%plot(t(1:end-1),ac,'.')
%title('Autocorrelation function')
%xlabel('$t  [ms]$','Interpreter','latex','FontSize',16)
%ylabel('$C(t)  [V^{2}]$','Interpreter','latex','FontSize',16 )

%figure(2)
%semilogy(t(1:end-1),ac,'.')
%title('Autocorrelation function LOGY')
%xlabel('$t  [ms]$','Interpreter','latex','FontSize',16)
%ylabel('$C(t)  [V^{2}]$','Interpreter','latex','FontSize',16 )




%% fit with exponential decay
%%%%% instead of doing an exponential fit, I do a linear fit on (t,log(ac)). 
%%%%% If data are exponential decay, this should give a line in semilogy,
%%%%% tau is the characteristic  time of the exponential 

rangemax=13;% here we set the data that you want to fit. Depends on the angle
[p,s]= polyfit((1:rangemax), log(ac(1:rangemax)),1);
tau= -1/p(1)   %%% characteristic time of exponential decay

%%%% calculate error on p(1)
ste = sqrt(diag(inv(s.R)*inv(s.R')).*s.normr.^2./s.df)



%%%% plot the autocorrelation function with the fit

figure(3);
semilogy(1:rangemax, ac(1:rangemax),'ro'); hold on;
semilogy(1:rangemax, exp(polyval(p,1:rangemax)),'b-','LineWidth',2);

xlabel('$t  [ms]$','Interpreter','latex','FontSize',16)
ylabel('$C(t)  [V^{2}]$','Interpreter','latex','FontSize',16 )
h = legend('Exp data', 'fit $ -e^{t/\tau}$');
set(h,'Interpreter','latex','FontSize',14);
title('Correlation function semilogy scale ','Interpreter','latex','FontSize',14);



figure(4);
rangemax= rangemax + 100;
plot(1:rangemax, ac(1:rangemax),'ro'); hold on;
plot(1:rangemax, exp(polyval(p,1:rangemax)),'b-','LineWidth',2);
xlabel('$t  [ms]$','Interpreter','latex','FontSize',16);
ylabel('$C(t)  [V^{2}]$','Interpreter','latex','FontSize',16 );
h = legend('Exp data', 'fit $ -e^{t/\tau}$');
set(h,'Interpreter','latex','FontSize',14);
title('Correlation function linear scale','Interpreter','latex','FontSize',14);






